### PR TITLE
In Memory Cache For Case Search

### PR DESCRIPTION
**Pull request by shubham1g5**
Alternative implementation for https://github.com/dimagi/formplayer/pull/850

https://dimagi-dev.atlassian.net/browse/USH-581

cross-request: https://github.com/dimagi/commcare-core/pull/979

Both core and this PR should be reverted together in case of rollbacks

### CODE REVIEW

File: src/main/java/org/commcare/formplayer/services/CaseSearchHelper.java
Comment (shubham1g5) [inline src/main/java/org/commcare/formplayer/services/CaseSearchHelper.java:36]: Hey @snopoke, I am a bit stuck while implementing caching for Case Search. This method is marked as `@Cacheable` though somehow it's getting executed everytime for the same cache key aka same username and uri passed. Would you be able to take a quick look and see if I am missing something obvious in this implementation.

Comment (snopoke) [reply src/main/java/org/commcare/formplayer/services/CaseSearchHelper.java:36]: are you testing locally? you need to enable the cache backend: `spring.cache.type=caffeine` (in `application.config`)

Comment (snopoke) [reply src/main/java/org/commcare/formplayer/services/CaseSearchHelper.java:36]: I think you should also add `case_search` to the list of cache names: https://github.com/dimagi/formplayer/blob/96caf7cb59a09aeb42ba687c36d5ec7364088d11/src/main/resources/application.properties#L33

File: src/main/java/org/commcare/formplayer/services/CaseSearchHelper.java
Comment (snopoke) [inline src/main/java/org/commcare/formplayer/services/CaseSearchHelper.java:46]: we should be able to do this with the cache configuration `spring.cache.caffeine.spec=expireAfterWrite=30m`. We could apply that globally to all caches or we could configure the caches separately (https://www.baeldung.com/spring-multiple-cache-managers)

File: src/main/java/org/commcare/formplayer/services/CaseSearchHelper.java
Comment (ctsims) [inline src/main/java/org/commcare/formplayer/services/CaseSearchHelper.java:41]: One thing about the `ExternalDataInstance`: I *don't* think it's safe, unfortunately, for us to reuse the raw java object, since it's not built for concurrency. 

@snopoke Is there a way to manually tell the system HOW to cache the object? It's fairly easy to marshall/unmarshall into a binary format which acts as a deep copy, but it would be nice to only need to do that when the object is retrieved from the cache

**Review by ctsims — COMMENTED** [review]
A few comments on the in process work.

File: src/main/java/org/commcare/formplayer/services/CaseSearchHelper.java
Comment (snopoke) [reply src/main/java/org/commcare/formplayer/services/CaseSearchHelper.java:41]: I don't know of a way to do this with the annotations, it might be easier in this case to use the raw cache API:

```java
@Autowired
private CacheManager cacheManager;

...

cache = cacheManager.getCache("case_search")
value = cache.get(key, valueLoader)
...
cache.put(key, serializedValue)
```

File: src/main/java/org/commcare/formplayer/services/CaseSearchHelper.java
Comment (shubham1g5) [reply src/main/java/org/commcare/formplayer/services/CaseSearchHelper.java:36]: that was it! I was missing both of these params. Thanks a lot.

File: src/main/java/org/commcare/formplayer/services/CaseSearchHelper.java
Comment (shubham1g5) [reply src/main/java/org/commcare/formplayer/services/CaseSearchHelper.java:46]: @snopoke Lookds like configuring a cache manager in Java code will override the `.properties` config and will enable the cache by default without having the ability to disable the cache using `.properties` file. Therefore I am going to lean towards applying a global config for all caches to `expireAfterWrite` after 5 mins.

Comment (snopoke) [reply src/main/java/org/commcare/formplayer/services/CaseSearchHelper.java:46]: I think that's fine for the current use cases. The ability to disable it was only intended for initial rollout so it would also be fine to have them permanently configured. Or if we really want we can take an approach like this: https://medium.com/@d.lopez.j/configuring-multiple-ttl-caches-in-spring-boot-dinamically-75f4aa6809f3

File: src/main/java/org/commcare/formplayer/services/CaseSearchHelper.java
Comment (shubham1g5) [reply src/main/java/org/commcare/formplayer/services/CaseSearchHelper.java:33]: @ctsims I am having trouble serializing `ExternalDataInstance` correctly, It seems to skip the [root node](https://github.com/dimagi/commcare-core/blob/master/src/main/java/org/javarosa/core/model/instance/ExternalDataInstance.java#L19) during [serialization](https://github.com/dimagi/commcare-core/blob/master/src/main/java/org/javarosa/core/model/instance/ExternalDataInstance.java#L99) today as root type `AbstractTreeElement` is an interface here. I am not sure what's the best way forward to serialize this. 

Also I did survey ExternalDataInstance and it's containing objects and they all are completely immutable.

Comment (ctsims) [reply src/main/java/org/commcare/formplayer/services/CaseSearchHelper.java:33]: Ah, I see what's going on. I mistook what the serialization of `ExternalDataInstance` actually accomplished.

It looks like `ExternalDataInstance`'s serialization is used to store **a placeholder** for a data instance that is connected up at runtime and that when we serialize concrete DOM's (like an inline instance in a form) we end up parsing them as [FormInstance](https://github.com/dimagi/commcare-core/blob/master/src/main/java/org/javarosa/xform/parse/XFormParser.java#L541) objects.

Options moving forward:

*One:* You could instead switch to serializing the [TreeEleement Root](https://github.com/dimagi/commcare-core/blob/master/src/main/java/org/commcare/session/RemoteQuerySessionManager.java#L131) itself, since that's how External Data Instance is actually created. `TreeElement`s should be reliably portable. 

*Two:* Given the added complexity of the serialized deep copy, I think it sounds fine to start with the assumption that we'll cache the in-memory version. It does look like the underlying structure is coded with a `TreeElement` DOM which is mutable and has [at least one](https://github.com/dimagi/commcare-core/blob/b47f46d332abd402ba7833dd6617f03306272bf9/src/main/java/org/javarosa/core/model/instance/TreeElement.java#L81) synchronicity concern, but since the DOM is used immutably and doesn't change after init I think it won't be an issue.

File: build.gradle
Comment (snopoke) [inline build.gradle:98]: why was it necessary to change this?

File: src/main/java/org/commcare/formplayer/application/MenuController.java
Comment (snopoke) [inline src/main/java/org/commcare/formplayer/application/MenuController.java:35]: why not just make the autowired bean in the AbstractBaseController `protected`?

File: src/main/java/org/commcare/formplayer/application/WebAppContext.java
Comment (snopoke) [inline src/main/java/org/commcare/formplayer/application/WebAppContext.java:250]: you could just tag the class with `@Component` and let the auto-discovery pick it up

File: src/main/java/org/commcare/formplayer/services/QueryRequester.java
Comment (snopoke) [inline src/main/java/org/commcare/formplayer/services/QueryRequester.java:15]: this class can be removed

File: build.gradle
Comment (shubham1g5) [reply build.gradle]: reverted this (was experimenting with multiple cache managers before and is not required anymore)

File: src/main/java/org/commcare/formplayer/services/CaseSearchHelper.java
Comment (ctsims) [inline src/main/java/org/commcare/formplayer/services/CaseSearchHelper.java:40]: Nice! Glad to see you found a way through, here.

File: src/main/java/org/commcare/formplayer/util/SerializationUtil.java
Comment (ctsims) [inline src/main/java/org/commcare/formplayer/util/SerializationUtil.java:34]: I think the ExtUtil class has an equivalent method?

File: src/main/resources/application.properties
Comment (ctsims) [inline src/main/resources/application.properties:31]: @snopoke Is it ok / good practice for us to use the same timings and such for all of these, or should be be trying to tune per-cache?

**Review by ctsims — APPROVED** [review]
This seems like a good approach to me, although I'm curious to get @snopoke 's inputs on the caching elements specifically

File: src/main/resources/application.properties
Comment (snopoke) [reply src/main/resources/application.properties:31]: This is fine unless the tuning requirements are vastly different. We did discuss this briefly and decided that apply the same expiration to the session caching wouldn't be an issue. Now that I look at it again I think it would be better to have them separate to avoid expiring the session cache in the middle of a form session. I don't think it's worth blocking the PR. Once this is merged I'll implement this pattern: https://medium.com/@d.lopez.j/configuring-multiple-ttl-caches-in-spring-boot-dinamically-75f4aa6809f3

**Pull request merged**
